In [1]:
import numpy as np
import pandas as pd

inf_ = np.inf
maxt_ = 5000.0

## Read in the Data and clean

In [2]:
train = pd.read_csv('../data/train.csv', dtype={'is_duplicate': np.float32})
test = pd.read_csv('../data/test.csv')

In [3]:
X = pd.read_csv('features.csv', header=None)

In [4]:
y = train.loc[:,'is_duplicate'].as_matrix()

In [5]:
bads = X[X.loc[:,1] >= 100].index
X.loc[bads,1] = 100.000

"""""""""ALTERNATE FEATURE LOAD"""""""""""""""

In [10]:
X = pd.read_csv('kfwmd_1.csv')
inf_ = max(X['kfwmd'])
bads = X[X['kfwmd'] == inf_].index
X.loc[bads,'kfwmd'] = maxt_
max(X['kfwmd'])

5000.0

In [8]:
X = pd.read_csv('unqwmd_1.csv')
inf_ = max(X['kfwmd'])
bads = X[X['kfwmd'] == inf_].index
X.loc[bads,'kfwmd'] = maxt_
max(X['kfwmd'])

5000.0

In [10]:
X = pd.read_csv('vnwmd_1.csv')
inf_ = max(X['kfwmd'])
bads = X[X['kfwmd'] == inf_].index
X.loc[bads,'kfwmd'] = maxt_
max(X['kfwmd'])

5000.0

In [12]:
X = pd.read_csv('entwmd_1.csv')
inf_ = max(X['kfwmd'])
bads = X[X['kfwmd'] == inf_].index
X.loc[bads,'kfwmd'] = maxt_
max(X['kfwmd'])

5000.0

In [14]:
X = pd.read_csv('pos.csv', index_col= 0)
inf_ = np.inf
X = X.replace(inf_, maxt_)

# Combining the feature sets

In [12]:
def clean(X):
    return X.replace(inf_, maxt_)

In [ ]:
X_5 = pd.read_csv('featz.csv', header=None)

In [102]:
X_1 = clean(pd.read_csv('entwmd.csv'))['kfwmd'].as_matrix()
X_2 = clean(pd.read_csv('vnwmd.csv'))['kfwmd'].as_matrix()
X_3 = clean(pd.read_csv('unqwmd.csv'))['kfwmd'].as_matrix()
X_4 = clean(pd.read_csv('kfwmd.csv'))['kfwmd'].as_matrix()

bads = X_5[X_5.loc[:,1] >= 100].index
X_5.loc[bads,1] = 100.000

X = np.matrix([X_1,X_2,X_3,X_4,X_5[0].as_matrix(),X_5[1].as_matrix() ])
X = X.T
X.shape

(404290, 6)

In [90]:
X_1 = clean(pd.read_csv('entwmd.csv'))['kfwmd'].as_matrix()
X_2 = clean(pd.read_csv('vnwmd.csv'))['kfwmd'].as_matrix()
X_3 = clean(pd.read_csv('unqwmd.csv'))['kfwmd'].as_matrix()
X_4 = clean(pd.read_csv('kfwmd.csv'))['kfwmd'].as_matrix()

bads = X_5[X_5.loc[:,1] >= 100].index
X_5.loc[bads,1] = 100.000

X = np.matrix([X_1,X_2,X_3,X_4]).T
X = np.hstack((X,X_5))
X.shape

(404290, 306)

# on inspection of features

In [5]:
X

0            1           2            3            4   \
0       5000.000000     0.000000    5.660510  5000.000000     0.000000   
1       5000.000000     0.000000   34.407731     0.000000     0.000000   
2       5000.000000     0.000000   24.232860     0.000000     0.000000   
3          6.168255     0.000000   62.953509     0.000000  2500.000000   
4          3.879514     0.000000   63.825153     0.000000     0.000000   
5       5000.000000     0.000000   37.530353     6.137929     0.000000   
6       5000.000000     0.000000   49.806744     0.000000     0.000000   
7       5000.000000     0.000000    6.040206     0.000000     0.000000   
8       5000.000000    10.726514    9.876929  5000.000000     5.363257   
9       5000.000000     0.000000   19.699194     0.000000     0.000000   
10      5000.000000     0.000000   45.468284     0.000000     0.000000   
11      5000.000000     0.000000   12.863821     0.000000     5.916970   
12      5000.000000     0.000000    0.000000  5000.000000     0.000000   
13         0.000000     0.000000    5.181824  5000.000000     0.000000   
14      5000.000000     0.000000    0.000000  5000.000000     0.000000   
15      5000.000000     0.000000   64.225083     0.000000     0.000000   
16      5000.000000     0.000000    0.000000  5000.000000     0.000000   
17      5000.000000     0.000000   13.528172     0.000000     0.000000   
18      5000.000000     0.000000   33.626676     0.000000     0.000000   
19      5000.000000     0.000000    6.414459     0.000000     0.000000   
20      5000.000000     0.000000   17.305620    13.214126     6.869703   
21      5000.000000     0.000000   16.503767     0.000000     0.000000   
22      5000.000000     0.000000    6.092464     0.000000     0.000000   
23         5.331227     0.000000   47.670527     0.000000     0.000000   
24      5000.000000     0.000000   50.716319     0.000000     0.000000   
25      5000.000000     0.000000   18.390491  5000.000000     0.000000   
26      5000.000000     0.000000    6.509964  5000.000000     0.000000   
27      5000.000000     0.000000   22.051001     0.000000     0.000000   
28      5000.000000     0.000000    7.875324     0.000000     0.000000   
29         5.024429     0.000000   34.168512     0.000000     9.616929   
...             ...          ...         ...          ...          ...   
404260     0.000000    14.501110    6.312711    10.561130     0.000000   
404261  5000.000000     0.000000   12.645326     0.000000     0.000000   
404262  5000.000000     0.000000   13.020127     0.000000     7.663606   
404263  5000.000000     0.000000   21.277785     0.000000     0.000000   
404264  5000.000000     0.000000   60.465572     0.000000     0.000000   
404265  5000.000000     0.000000   13.847070     0.000000     0.000000   
404266     5.269974     0.000000  124.772166     0.000000     0.000000   
404267  5000.000000     0.000000   11.290917     0.000000     6.772803   
404268     1.960195    13.814607   48.321974     5.188987     6.907303   
404269  5000.000000     0.000000   15.650791     0.000000     0.000000   
404270  5000.000000     0.000000    0.000000  5000.000000     0.000000   
404271     4.462813     0.000000   44.524432     6.137929     0.000000   
404272  5000.000000     0.000000   53.517858     0.000000     0.000000   
404273  5000.000000     0.000000   32.722558     0.000000     0.000000   
404274  5000.000000     0.000000   28.473808     0.000000     0.000000   
404275  5000.000000     0.000000   24.978921     0.000000     0.000000   
404276  5000.000000  5000.000000    0.000000  5000.000000  2500.000000   
404277  5000.000000  5000.000000   10.637012  5000.000000  2500.000000   
404278  5000.000000     0.000000   14.424731     0.000000     0.000000   
404279     5.269974     0.000000   46.625725     8.249516     0.000000   
404280  5000.000000     0.000000    6.605224  5000.000000     0.000000   
404281  5000.000000     0.000000   20.130139     0.000000     0.000000   
404282  5000.000000  

In [6]:
X = X.replace(inf_,maxt_)

## Evaluate on various Methods
- Logistic
- Linear
- SGD
- Ridge
- LassoLars

In [40]:
from sklearn.linear_model import LogisticRegression, SGDRegressor, LinearRegression

from sklearn.model_selection import cross_val_score
regs = [
    LogisticRegression(),
    #SGDRegressor(),
    #LinearRegression(),
    #RandomForestRegressor()
]
for clf in regs:
    print clf
    scores = cross_val_score(clf, X, y, cv=3, scoring='neg_log_loss')
    print "Log Loss: %0.5f" % scores.mean()

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)
Log Loss: -0.57002


In [48]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss

def test_est(est):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)
    predy = est.fit(X_train,y_train).predict(X_test)
    print "Log Loss: %0.5f" % log_loss(y_test, predy)

In [49]:
from sklearn.linear_model import LinearRegression

test_est(LinearRegression())

Log Loss: 0.75071


In [50]:
from sklearn.linear_model import Ridge

test_est(Ridge(alpha=1.0))

Log Loss: 0.75070


In [52]:
from sklearn.linear_model import LassoLars

test_est(LassoLars())

Log Loss: 0.65906


In [45]:
from sklearn.linear_model import SGDRegressor
test_est(SGDRegressor(loss='huber'))
test_est(SGDRegressor(loss='squared_loss'))
test_est(SGDRegressor(loss='squared_epsilon_insensitive'))
test_est(SGDRegressor(loss='epsilon_insensitive'))

Log Loss: 10.11853
Log Loss: 14.28055
Log Loss: 19.17337
Log Loss: 16.10024


In [11]:
from sklearn.linear_model import SGDRegressor
from sklearn.linear_model import LassoLars
from sklearn.linear_model import Ridge
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss
def test_est(est):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=0)
    predy = est.fit(X_train,y_train).predict(X_test)
    print "Log Loss: %0.5f" % log_loss(y_test, predy)
from sklearn.linear_model import LogisticRegression

from sklearn.model_selection import cross_val_score

print "Logistic"
scores = cross_val_score( LogisticRegression(), X, y, cv=3, scoring='neg_log_loss')
print "Log Loss: %0.5f" % scores.mean()

print "Linear"
test_est(LinearRegression())

print "Ridge"
test_est(Ridge(alpha=1.0))

print "LassoLars"
test_est(LassoLars())


print "SGDs"
test_est(SGDRegressor(loss='huber'))
test_est(SGDRegressor(loss='squared_loss'))
test_est(SGDRegressor(loss='squared_epsilon_insensitive'))
test_est(SGDRegressor(loss='epsilon_insensitive'))
    

Logistic
Log Loss: -0.59030
Linear


/usr/local/lib/python2.7/site-packages/scipy/linalg/basic.py:1018: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
  warnings.warn(mesg, RuntimeWarning)


Log Loss: 0.64659
Ridge
Log Loss: 0.64659
LassoLars
Log Loss: 0.65906
SGDs
Log Loss: 12.78577
Log Loss: 21.75351
Log Loss: 12.78620
Log Loss: 12.78577


In [30]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import log_loss

log = LogisticRegression()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
predy = log.fit(X_train,y_train).predict_proba(X_test)

predy[:,1].shape, y_test.shape

((80858,), (80858,))

In [37]:
np.vstack((predy[:,1], y_test)).T

array([[ 0.42462291,  1.        ],
       [ 0.34824287,  1.        ],
       [ 0.24498837,  0.        ],
       ..., 
       [ 0.52251958,  1.        ],
       [ 0.40346384,  0.        ],
       [ 0.52492596,  1.        ]])

In [36]:

np.savetxt('probs.csv',np.vstack((predy[:,1], y_test)).T,delimiter=',')
print "Log Loss: %0.5f" % log_loss(y_test, predy)

Log Loss: 0.58749


# Results
## my_wmd
Log Loss: 0.64659  
Ridge  
Log Loss: 0.64659  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 12.78577  
Log Loss: 12.78577  
Log Loss: 12.78577  
Log Loss: 12.78577  

## unique wmd
Logistic  
Log Loss: -0.67535  
Linear  
Log Loss: 0.65143  
Ridge  
Log Loss: 0.65143  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 12.78577  
Log Loss: 12.78470  
Log Loss: 12.78599  
Log Loss: 12.78577  

## verbs and nouns comb
Logistic  
Log Loss: -0.67462  
Linear  
Log Loss: 0.65875  
Ridge  
Log Loss: 0.65875  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 21.75308  
Log Loss: 21.75351  
Log Loss: 12.78727  
Log Loss: 12.78577  

# entities
Logistic  
Log Loss: -0.67568  
Linear  
Log Loss: 0.65733  
Ridge  
Log Loss: 0.65733  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 21.75351  
Log Loss: 12.81439  
Log Loss: 21.73386  
Log Loss: 21.70032  

# pos wtd
Logistic  
Log Loss: -0.65728  
Linear  
Log Loss: 0.65766  
Ridge  
Log Loss: 0.65766  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 17.49136  
Log Loss: 21.75351  
Log Loss: 21.40920  
Log Loss: 21.39288  

# combined (w/o cc)
Logistic  
Log Loss: -0.57243  
Linear  
Log Loss: 0.75071  
Ridge  
Log Loss: 0.75070  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 9.79779  
Log Loss: 18.78757  
Log Loss: 19.62294  
Log Loss: 18.68083  

# Combined Correct
Logistic  
Log Loss: -0.57723  
Linear  
Log Loss: 0.74849  
Ridge  
Log Loss: 0.74849  
LassoLars  
Log Loss: 0.65906  
SGDs  
Log Loss: 13.03172  
Log Loss: 21.74988  
Log Loss: 21.66294  
Log Loss: 21.42330  